### PR TITLE
chore: add eslint unicorn

### PR DESCRIPTION
**Pull request by Shinigami92**
Based on https://github.com/faker-js/faker/pull/2284#discussion_r1333960666

This PR will make @fisker really happy :heart:

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -18,14 +18,15 @@ module.exports = defineConfig({
...
+    'plugin:unicorn/recommended',
   ],
   parser: '@typescript-eslint/parser',
   parserOptions: {
     project: ['./tsconfig.json'],
     sourceType: 'module',
     warnOnUnsupportedTypeScriptVersion: false,
   },
-  plugins: ['@typescript-eslint', 'prettier', 'deprecation'],
+  plugins: ['@typescript-eslint', 'prettier', 'deprecation', 'unicorn'],
```
Comment (ST-DDT) [inline .eslintrc.js:29]: This is optional/already covered by the extends config.

File: .eslintrc.js [hunk 2]
```diff
@@ -36,6 +37,19 @@ module.exports = defineConfig({
 
     'deprecation/deprecation': 'error',
 
+    'unicorn/no-array-reduce': 'off',
+    'unicorn/no-null': 'off',
+    'unicorn/prevent-abbreviations': [
+      'error',
+      {
+        allowList: {
+          args: true,
```
Comment (ST-DDT) [inline .eslintrc.js:65]: Why args but not arg?

File: test/modules/helpers.spec.ts [hunk 3]
```diff
@@ -48,18 +48,18 @@ describe('helpers', () => {
       t.it('with static string', 'Hello World!')
         .it('with static RegExp', /Hello World!/)
         .it('with dynamic string', '[A-D0-9]-[A-D0-9]-A')
-        .it('with dynamic RegExp', /[A-D0-9]-[A-D0-9]-A/)
+        .it('with dynamic RegExp', /(?:[\dA-D]-){2}A/)
```
Comment (ST-DDT) [inline test/modules/helpers.spec.ts]: These are tests, we might need both variants. IIRC the new variant isnt supported yet.

File: test/modules/helpers.spec.ts [hunk 4]
```diff
@@ -461,7 +461,7 @@ describe('helpers', () => {
           const input = Array.from({ length: 2 }, (_, i) => i);
           const occurrences = Array.from({ length: 2 }, () => 0);
 
-          for (let i = 0; i < 1000; i++) {
+          for (let index = 0; index < 1000; index++) {
```
Comment (ST-DDT) [inline test/modules/helpers.spec.ts]: This isnt the index it is a counter.

File: test/modules/helpers.spec.ts [hunk 5]
```diff
@@ -679,18 +671,18 @@ describe('helpers', () => {
         });
 
         it('deals with RegExp object', () => {
-          const string = faker.helpers.fromRegExp(/[A-D0-9]{4}-[A-D0-9]{4}/);
-          expect(string).toMatch(/^[A-D0-9]{4}-[A-D0-9]{4}$/);
+          const string = faker.helpers.fromRegExp(/[\dA-D]{4}-[\dA-D]{4}/);
```
Comment (ST-DDT) [inline test/modules/helpers.spec.ts]: Maybe this isnt supported yet.

File: .eslintrc.js [hunk 6]
```diff
@@ -36,6 +37,20 @@ module.exports = defineConfig({
...
+          fn: true,
+          obj: true,
```
Comment (ST-DDT) [inline .eslintrc.js:67]: Why obj, but not str/num?

File: test/modules/color.spec.ts [hunk 7]
```diff
@@ -44,8 +44,8 @@ describe('color', () => {
 
       describe(`cssSupportedFunction()`, () => {
         it('should return random css supported color function from css functions array', () => {
-          const func = faker.color.cssSupportedFunction();
-          expect(Object.values(CssFunction)).toContain(func);
+          const supportedFunction = faker.color.cssSupportedFunction();
```
Comment (ST-DDT) [inline test/modules/color.spec.ts:47]: Maybe name it actual same as in our other tests?

File: test/modules/date.spec.ts [hunk 8]
```diff
@@ -358,11 +358,11 @@ describe('date', () => {
             expect(dates.length).lessThanOrEqual(5);
 
             expect(dates[0]).greaterThan(from);
-            for (let i = 1; i < dates.length; i++) {
-              expect(dates[i]).greaterThan(dates[i - 1]);
+            for (let index = 1; index < dates.length; index++) {
```
Comment (ST-DDT) [inline test/modules/date.spec.ts:361]: This seems unnessarily verbose.

**Comment by ST-DDT** [comment]
Note to self. Check all the regexes again.

**Comment by Shinigami92** [comment]
@ST-DDT I highly suggest you start reviewing this when I'm actually done with it
There are more then 800 errors I need to manually go through
I'm already sitting on this more than an hour and I make a break now, playing Overwatch

File: test/modules/helpers.spec.ts [hunk 9]
```diff
@@ -475,10 +475,10 @@ describe('helpers', () => {
         it.each([10, 100, 1000])(
           'should return each element with a somewhat equal distribution with %s elements',
           (length) => {
-            const input = Array.from({ length }, (_, i) => i % 10);
+            const input = Array.from({ length }, (_, index) => index % 10);
             const occurrences = Array.from({ length: 10 }, () => 0);
 
-            for (let i = 0; i < 1000; i++) {
+            for (let index = 0; index < 1000; index++) {
```
Comment (ST-DDT) [inline test/modules/helpers.spec.ts:483]: This isnt an index.

File: test/modules/image.spec.ts [hunk 10]
```diff
@@ -485,7 +485,7 @@ describe('image', () => {
 
       expect(imageUrl).toBeTypeOf('string');
       expect(imageUrl).toMatch(
-        /^https\:\/\/via\.placeholder\.com\/\d+x\d+\/[0-9a-fA-F]{6}\/[0-9a-fA-F]{6}\.[a-z]{3,4}\?text=.+$/
+        /^ht{2}ps:\/{2}via\.placeholder\.com\/\d+x\d+(?:\/[\dA-Fa-f]{6}){2}\.[a-z]{3,4}\?text=.+$/
```
Comment (ST-DDT) [inline test/modules/image.spec.ts:488]: No. `ht{2}ps:\/{2}` :no_good:

File: test/modules/internet.spec.ts [hunk 11]
```diff
@@ -297,7 +297,7 @@ describe('internet', () => {
           const [prefix, suffix] = email.split('@');
 
           expect(prefix).toMatch(
-            /^Mike((\d{1,2})|([.!#$%&'*+-/=?^_`{|}~]Smith\d{1,2})|([.!#$%&'*+-/=?^_`{|}~]Smith))/
+            /^Mike((\d{1,2})|([!#$%&'*+-/=?^_`{|}~]Smith\d{1,2})|([!#$%&'*+-/=?^_`{|}~]Smith))/
```
Comment (ST-DDT) [inline test/modules/internet.spec.ts:300]: Note to self: Where did the dot go?

File: test/modules/lorem.spec.ts [hunk 12]
```diff
@@ -78,15 +78,15 @@ describe('lorem', () => {
 
         it.each(times(25))(
           'should return %i random values from words array',
-          (num) => {
-            const actual = faker.lorem.words(num);
+          (number) => {
```
Comment (ST-DDT) [inline test/modules/lorem.spec.ts:81]: Why number here and number_ at other locations IIRC?

File: test/support/seeded-runs.ts [hunk 13]
```diff
@@ -142,7 +142,7 @@ class TestGenerator<
     repetitions: number = 1
   ): void {
     this.setup();
-    for (let i = 0; i < repetitions; i++) {
+    for (let index = 0; index < repetitions; index++) {
```
Comment (ST-DDT) [inline test/support/seeded-runs.ts:145]: This isnt an index.

**Comment by ST-DDT** [comment]
> @ST-DDT I highly suggest you start reviewing this when I'm actually done with it There are more then 800 errors I need to manually go through I'm already sitting on this more than an hour and I make a break now, playing Overwatch

I'm not sure I want to review 800 changes in one PR.

Correction: The current number of addressed (types of) issues already exceeds what I'm willing to review in a single PR.

**Review by ST-DDT — CHANGES_REQUESTED** [review]
This PR is not reviewable due to size/different kind of changes.

**Comment by Shinigami92** [comment]
As @ST-DDT already pointed out, I need to split this into several PRs, to reduce the amount of changed files in the same PR.

This PR now reflects the approximate change when everything is merged individually.

While working on this PR, I at least learned some stuff I can reuse to work on the individual PRs.
And maybe this PR gets so much reduces in diff count, that it can then be used to check the rest that is left over.

**Comment by Shinigami92** [comment]
I switched from `on hold` to `do not merge`, because `on hold` is for that the work is not going forward, which is effectively wrong in this case

**Comment by ST-DDT** [comment]
This has been superseded by #2418 and individual PRs per rule.

Since Shinigami92 is no longer actively working on this. I will close this because of the merge conflicts.
The diff is visible here forever and the branch can be restored if needed as well.

**Pull request closed without merging**